### PR TITLE
sim: simsensors: support external simulator

### PR DESCRIPTION
**Pull request by mlyle**

### CODE REVIEW

**Comment by mlyle** [comment]
jenkins, build artifacts please

**Comment by mlyle** [comment]
jenkins, test this please

File: flight/Modules/Sensors/simulated/simsensors.c [hunk 1]
```diff
@@ -241,6 +241,7 @@ static void SensorsTask(void *parameters)
 
 static void simulateYasim()
 {
+#if !(defined(_WIN32) || defined(WIN32) || defined(__MINGW32__))
```
Comment (tracernz) [inline flight/Modules/Sensors/simulated/simsensors.c]: I see this combo a lot. We should probably have a macro somewhere so it's maintainable (and actually, our Makefiles already determine this right?).

Comment (mlyle) [reply flight/Modules/Sensors/simulated/simsensors.c]: Yes-- probably should have a different sub-architecture though (like we have common/stm32/stm32f4xx .. could have common/posix/posixwin/ or whatever, with associated defines etc like firmware-arches does

File: flight/Modules/Sensors/simulated/simsensors.c [hunk 2]
```diff
@@ -216,119 +222,268 @@ static void SensorsTask(void *parameters)
...
+				simulateYasim();
+				break;
 		}
 
 		static uint32_t tm = 0;
 
-		PIOS_Thread_Sleep_Until(&tm, 2);
+		if (use_yasim) {
+			// 200Hz.
+			PIOS_Thread_Sleep_Until(&tm, 5);
```
Comment (tracernz) [inline flight/Modules/Sensors/simulated/simsensors.c]: `PIOS_Thread_Sleep_Until(&tm, 1000 / sens_rate)`?

File: flight/Modules/Sensors/simulated/simsensors.c [hunk 3]
```diff
@@ -85,7 +86,9 @@ static float rand_gauss();
 
 static bool use_real_sensors;
 
-enum sensor_sim_type {CONSTANT, MODEL_AGNOSTIC, MODEL_QUADCOPTER, MODEL_AIRPLANE, MODEL_CAR} sensor_sim_type;
+extern bool use_yasim;
```
Comment (tracernz) [inline flight/Modules/Sensors/simulated/simsensors.c]: Would be nice to pass this through nicely somehow. PIOS_Sensors seems like the right place.. but not specific to sim currently :/.

File: flight/Modules/Sensors/simulated/simsensors.c [hunk 4]
```diff
@@ -216,119 +222,268 @@ static void SensorsTask(void *parameters)
...
-	// Apply bias correction to the gyros
-	GyrosBiasData gyrosBias;
-	GyrosBiasGet(&gyrosBias);
-	gyrosData.x += gyrosBias.x;
-	gyrosData.y += gyrosBias.y;
-	gyrosData.z += gyrosBias.z;
+	    float roll, pitch, yaw, throttle;
 
-	GyrosSet(&gyrosData);
+	    float resv[8];
```
Comment (tracernz) [inline flight/Modules/Sensors/simulated/simsensors.c]: Reserved?

File: flight/Modules/Sensors/simulated/simsensors.c [hunk 5]
```diff
@@ -216,119 +222,268 @@ static void SensorsTask(void *parameters)
...
+
+		// Run the offset compensation algorithm from the firmware
+		magOffsetEstimation(&mag);
+
+		MagnetometerSet(&mag);
+		last_mag_time = PIOS_DELAY_GetRaw();
+	}
+
+	static float baro_offset = 0;
+	if(baro_offset == 0) {
```
Comment (tracernz) [inline flight/Modules/Sensors/simulated/simsensors.c]: space

File: flight/Modules/Sensors/simulated/simsensors.c [hunk 6]
```diff
@@ -216,119 +222,268 @@ static void SensorsTask(void *parameters)
...
+		homeLocation.Be[0] = 100;
+		homeLocation.Be[1] = 0;
+		homeLocation.Be[2] = 400;
+		homeLocation.Set = HOMELOCATION_SET_TRUE;
+
+		HomeLocationSet(&homeLocation);
+	}
+
+	uint32_t last_mag_time = 0;
+	if(PIOS_DELAY_DiffuS(last_mag_time) / 1.0e6 > MAG_PERIOD) {
```
Comment (tracernz) [inline flight/Modules/Sensors/simulated/simsensors.c]: space

File: flight/Modules/Sensors/simulated/simsensors.c [hunk 7]
```diff
@@ -216,119 +222,268 @@ static void SensorsTask(void *parameters)
...
+		// Hacky initialization
+		baro_offset = 50;// * rand_gauss();
+	} else {
+		// Very small drift process
+		baro_offset += rand_gauss() / 100;
+	}
+
+	// Update baro periodically
+	static uint32_t last_baro_time = 0;
+	if(PIOS_DELAY_DiffuS(last_baro_time) / 1.0e6 > BARO_PERIOD) {
```
Comment (tracernz) [inline flight/Modules/Sensors/simulated/simsensors.c]: space

File: flight/Modules/Sensors/simulated/simsensors.c [hunk 8]
```diff
@@ -216,119 +222,268 @@ static void SensorsTask(void *parameters)
...
+				perror("dup2STDIN");
+				exit(1);
+			}
+
+			for (int i = 3; i < 1024; i++) {
+				close(i);
+			}
+
+			execlp("yasim-svr", "yasim-svr",
+					"/home/mlyle/drhil/pa22-160-yasim.xml",
```
Comment (tracernz) [inline flight/Modules/Sensors/simulated/simsensors.c]: Hardcoded path

File: flight/Modules/Sensors/simulated/simsensors.c [hunk 9]
```diff
@@ -216,119 +222,268 @@ static void SensorsTask(void *parameters)
...
+	    double lat, lon, alt;
+
+	    float p, q, r;
+	    float acc[3];
+	    float vel[3];
+
+	    /* Provided only to "check" attitude solution */
+	    float roll, pitch, hdg;
+
+	    float resv[4];
```
Comment (tracernz) [inline flight/Modules/Sensors/simulated/simsensors.c]: The abbreviation only obfuscates IMO, and it's not used so no win in brevity.

**Review by tracernz — CHANGES_REQUESTED** [review]
Hardcoded path.

**Comment by mlyle** [comment]
Woops :P

File: flight/Modules/Sensors/simulated/simsensors.c (hunk 9 above)
Comment (mlyle) [reply flight/Modules/Sensors/simulated/simsensors.c]: It's the "standard" way to write reserved.  That or someprefix_reserved.

File: flight/Modules/Sensors/simulated/simsensors.c (hunk 3 above)
Comment (mlyle) [reply flight/Modules/Sensors/simulated/simsensors.c]: Yah.  It's like posix needs some kind of configuration registry for things like this--- ultimately moving away from the command line to a config file and adding a small API would help for this.

**Review by tracernz — APPROVED** [review]

**Comment by mlyle** [comment]
jenkins, test this please

**Pull request merged**